libs/langchain-community/src/retrievers/matryoshka_retrieval.ts

```diff
@@ -0,0 +1,128 @@
+import { CallbackManagerForRetrieverRun } from "@langchain/core/callbacks/manager";
+import { DocumentInterface } from "@langchain/core/documents";
+import {
+  VectorStore,
+  VectorStoreRetriever,
+  VectorStoreRetrieverInput,
+} from "@langchain/core/vectorstores";
+
+/**
+ * Type for options when adding a document to the VectorStore.
+ */
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type AddDocumentOptions = Record<string, any>;
+
+export interface MatryoshkaRetrievalFields<
+  Store extends VectorStore = VectorStore
+> extends Omit<VectorStoreRetrieverInput<Store>, "vectorStore"> {
+  /**
+   * The vector store using the larger embeddings model.
+   */
+  largeStore: Store;
+  /**
+   * The vector store using the smaller embeddings model.
+   */
+  smallStore: Store;
+  /**
+   * The number of documents to retrieve from the small store.
+   *
+   * @default 20
+   */
+  smallK?: number;
+  /**
+   * The number of documents to retrieve from the large store.
+   *
+   * @default 5
+   */
+  largeK?: number;
+  /**
+   * Optionally override the ID key used to filter the large store
+   * results.
+   *
+   * @default "id"
+   */
+  idKey?: string;
+}
+
+/**
+ * A retriever that uses two stores, a large and small (dictated by their dimensions)
+ * to retrieve documents.
+ *
+ * Based off of the "Matryoshka embeddings: faster OpenAI vector search using Adaptive Retrieval"
+ * blog post {@link https://supabase.com/blog/matryoshka-embeddings}.
+ *
+ * The MatryoshkaRetrieval retriever performs two semantic searches. The first uses the smaller
+ * store which returns many results, but is less accurate. Then, using the documents returned
+ * from the smaller store, the larger store is used to perform a more accurate search over the
```

Comment:
Could you go into a bit more detail as to how this works? What does "more accurate search" mean?